cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/PipelineSpecGenerator.java

```diff
@@ -383,8 +395,37 @@ private ValidatedPipeline validateConfig(ETLConfig config) {
         throw new IllegalArgumentException(
           String.format("Invalid connection %s. %s is not a stage.", connection, connection.getTo()));
       }
+
+      if (conditionStages.contains(connection.getFrom())) {
+        if (connection.getCondition() == null) {
+          String msg = String.format("For condition stage %s, the connection %s is not marked with either " +
+                                       "'true' or 'false'.", connection.getFrom(), connection);
+          throw new IllegalArgumentException(msg);
+        }
+
+        // check if connection from the condition node is marked as true or false multiple times
+        if (conditionBranch.containsKey(connection.getFrom())
+          && connection.getCondition().equals(conditionBranch.get(connection.getFrom()))) {
+          String msg = String.format("For condition stage '%s', more than one outgoing connections are marked as %s.",
+                                     connection.getFrom(), connection.getCondition());
+          throw new IllegalArgumentException(msg);
+        }
+        conditionBranch.put(connection.getFrom(), connection.getCondition());
+      }
+
+      if (outgoingConnections.get(connection.getFrom()) == null) {
+        outgoingConnections.put(connection.getFrom(), new HashSet<String>());
+      }
+      outgoingConnections.get(connection.getFrom()).add(connection.getTo());
+
+      if (incomingConnections.get(connection.getTo()) == null) {
+        incomingConnections.put(connection.getTo(), new HashSet<String>());
+      }
+      incomingConnections.get(connection.getTo()).add(connection.getFrom());
     }
 
+    validateConditionBranches(conditionStages, incomingConnections, outgoingConnections);
```

Comment:
this should also be checked near the end of this method, as we probably want the simpler types of errors to be thrown before this more complex one. For example, if they have a transform that isn't connected to anything, that should probably be raised first.